### PR TITLE
VCF/Genotype Functionality 

### PR DESCRIPTION
**Pull request by abhim00**

### CODE REVIEW

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#Contributors:
+#VCF.py
+#Kamil Slowikowski
```
[the remark this answers is not in the document]
Comment (akmorrow13) [reply mango-viz/bdgenomics/mango/io/VCFfile.py]: also remove from license and move to separate comment

File: mango-viz/bdgenomics/mango/io/genomicfile.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 #
 
 import modin.pandas as pd
-
+import pandas as opd
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/genomicfile.py]: where is pandas being used? can you set 

```
import modin.pandas as mpd
import pandas as pd

```

File: mango-viz/bdgenomics/mango/io/readers.py [hunk 3]
```diff
@@ -25,3 +26,6 @@ def read_bed(filepath_or_buffer,
     skiprows=None
  ):
    return BedFile.read(filepath_or_buffer, column_names, skiprows)
+
+def read_vcf(filepath_or_buffer):
+    return VCFFile.read(filepath_or_buffer)
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/readers.py]: add newline

File: mango-viz/bdgenomics/mango/io/test/test_vcf.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        assert(dataframe._pileup_visualization == "variantJson")
+        tracks=[Track(viz="variants", label="my variants", source=pileup.sources.DataFrameSource(dataframe))]
+        reads = pileup.PileupViewer(locus="chr22:10436-10564", reference="hg19", tracks=tracks)
+        assert(str(type(reads)) == '<class \'bdgenomics.mango.pileup.pileupViewer.PileupViewer\'>')
+
+
+    # Run tests
+if __name__ == '__main__':
+    unittest.main()
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_vcf.py]: add newline

File: mango-viz/bdgenomics/mango/pileup/sources.py [hunk 5]
```diff
@@ -38,7 +38,7 @@
     GA4GHVariantSource
     GA4GHFeatureSource
 """
-from bdgenomics.mango.io import *
+#from bdgenomics.mango.io import *
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/sources.py]: Can you move this to a separate PR?

File: mango-viz/bdgenomics/mango/io/test/test_vcf.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import unittest
+import json
+import pandas as pd
+from bdgenomics.mango.io import *
+import bdgenomics.mango.pileup as pileup
+from bdgenomics.mango.pileup.track import *
+from bdgenomics.mango.io.test import IOTestCase
+
+class VCFFileTest(IOTestCase):
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_vcf.py]: remove 1 blank line

File: mango-viz/bdgenomics/mango/io/test/test_vcf.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import unittest
+import json
+import pandas as pd
+from bdgenomics.mango.io import *
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_vcf.py]: remove *s if not needed

File: mango-viz/bdgenomics/mango/pileup/sources.py (hunk 5 above)
Comment (akmorrow13) [reply mango-viz/bdgenomics/mango/pileup/sources.py]: I already made a separate PR that does this. Make sure to rebase and this will go away.

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            # Return a simple DataFrame without splitting the INFO column.
+            df = cls.dataframe_lib.read_table(filename, compression=comp, skiprows=range(NUM_COMMENTS-1))
+            NUM_SAMPLES = len(df.columns) - len(VCF_HEADER)
+            df.columns = VCF_HEADER + [i for i in range(1, NUM_SAMPLES+1)]
+            return df       
+
+    @classmethod
+    def _getcallinformation(cls, row):
+        sample_column_names = row.index[9:]
+        call_information = '"calls":['
```
Comment (heuermh) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]: This doesn't look like the bdg-formats `Variant` or `Genotype` schema, or even similar to ADAM `VariantContext` or htsjdk `VariantContext`.  I assume the schema here is most closely aligned with the GA4GH schema, for sending to pileup?

Comment (abhim00) [reply mango-viz/bdgenomics/mango/io/VCFfile.py]: Yes, the schema is aligned with the ga4gh schema.

Comment (heuermh) [reply mango-viz/bdgenomics/mango/io/VCFfile.py]: 👍

**Comment by akmorrow13** [comment]
Nice @abhim00 glad you got it working!

**Comment by abhim00** [comment]
Thank You! @akmorrow13

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+class VCFFile(GenomicFile):
+
+    global VCF_HEADER
+    global NUM_SAMPLES
+    VCF_HEADER = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT']
+
+
+    @classmethod
+    def _read(cls, filename, large = True):
+        """Open an optionally gzipped VCF file and return a pandas.DataFrame with
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]: does this return a pandas dataframe, modin dataframe, or either?

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        each INFO field included as a column in the dataframe.
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]: Can you link to the VCF file specifications here?

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        
+        Borrowed from https://gist.github.com/slowkow/6215557
+
+        Note: Using large=False with large VCF files. It will be painfully slow.
+        :param filename:    An optionally gzipped VCF file.
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]: I am a little confused about the purpose of the large parameter. It looks like if large=False, nothing happens. Is this correct?

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        json_ga4gh = "{\"variants\":["
+        json_ga4gh += "{" + content_series.str.cat(sep="},{")[:-2]
+        json_ga4gh =  json_ga4gh + "}]}]}"
+        
+        return json_ga4gh
+        
+    
+    @classmethod
+    def _visualization(cls, df):
+        return 'variantJson'
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]: This is already defined by GA4GHVariantJson.name in sources.py, so you can pull the string from there

File: mango-viz/bdgenomics/mango/io/readers.py [hunk 13]
```diff
@@ -25,3 +26,7 @@ def read_bed(filepath_or_buffer,
     skiprows=None
  ):
    return BedFile.read(filepath_or_buffer, column_names, skiprows)
+
+def read_vcf(filepath_or_buffer):
+    return VCFFile.read(filepath_or_buffer)
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/readers.py]: add newline

File: mango-viz/bdgenomics/mango/io/test/__init__.py [hunk 14]
```diff
@@ -19,12 +19,13 @@
 
 import os
 import unittest
+from bdgenomics import mango
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/__init__.py]: do we need to import everything here?

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 15]
```diff
@@ -19,44 +19,25 @@
...
 
     def test_required_columns(self):
-        dataframe = read_bed(self.exampleFile("chr17.582500-594500.bed"))
-
+        dataframe = read_bed(self.exampleFile(filename))
         dataframe_columns = list(dataframe.columns)
         for name in ("chrom", "chromStart", "chromEnd"):
             assert(name in dataframe_columns)
 
-    def test_column_type(self):
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: why is this test removed?

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+        
+        Borrowed from https://gist.github.com/slowkow/6215557
+
+        Specification: https://samtools.github.io/hts-specs/VCFv4.2.pdf
+
+        Note: Using large=False with large VCF files. It will be painfully slow.
+        :param filename:    An optionally gzipped VCF file.
+        :param large:       Use this with large VCF files to skip the ## lines and
+                            leave the INFO fields unseparated as a single column.
+        """
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]: remove large param comment

File: mango-viz/bdgenomics/mango/io/genomicfile.py [hunk 17]
```diff
@@ -16,11 +16,10 @@
 # limitations under the License.
 #
 
-import modin.pandas as pd
-
+import modin.pandas as mpd
```
Comment (devin-petersohn) [inline mango-viz/bdgenomics/mango/io/genomicfile.py]:
```suggestion
import os

if os.environ.get("MANGO_USE_PANDAS", "False").title() == "False":
    import modin.pandas as pd
else:
    import pandas as pd
```

File: mango-viz/bdgenomics/mango/io/genomicfile.py [hunk 18]
```diff
@@ -16,11 +16,10 @@
 # limitations under the License.
 #
 
-import modin.pandas as pd
-
+import modin.pandas as mpd
 
 class GenomicFile(object):
-    dataframe_lib = pd
+    dataframe_lib = mpd
```
Comment (devin-petersohn) [inline mango-viz/bdgenomics/mango/io/genomicfile.py]:
```suggestion
    dataframe_lib = pd
```

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        
+        Borrowed from https://gist.github.com/slowkow/6215557
+
+        Specification: https://samtools.github.io/hts-specs/VCFv4.2.pdf
+        :param filename:    An optionally gzipped VCF file.
+       
+        """
+
+        # Set the proper argument if the file is compressed.
+        comp = 'gzip' if filename.endswith('.gz') else None
```
Comment (devin-petersohn) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]:
```suggestion
```

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        # Count how many comment lines should be skipped.
+        NUM_COMMENTS = cls._count_comments(filename)
+        # Return a simple DataFrame without splitting the INFO column.
+        df = cls.dataframe_lib.read_table(filename, compression=comp, skiprows=range(NUM_COMMENTS-1))
```
Comment (devin-petersohn) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]:
```suggestion
        df = cls.dataframe_lib.read_table(filename, skiprows=range(NUM_COMMENTS-1))
```

File: mango-viz/bdgenomics/mango/io/VCFfile.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        NUM_SAMPLES = len(df.columns) - len(VCF_HEADER)
+        df.columns = VCF_HEADER + [i for i in range(1, NUM_SAMPLES+1)]
```
Comment (devin-petersohn) [inline mango-viz/bdgenomics/mango/io/VCFfile.py]:
```suggestion
        df.columns = VCF_HEADER + list(range(1, NUM_SAMPLES+1))
```

**Comment by akmorrow13** [comment]
Jenkins, retest this please.
